src/natcap/invest/ndr/ndr_core.pyx

```diff
@@ -183,26 +166,23 @@ def ndr_eff_calculation(
                     global_col, global_row, effective_retention_nodata)
             else:
                 working_retention_eff = 0.0
-                outflow_weight_sum = 0
-                for i in range(8):
-                    outflow_weight = (flow_dir >> (i*4)) & 0xF
-                    if outflow_weight == 0:
-                        continue
-                    outflow_weight_sum += outflow_weight
-                    ds_col = col_offsets[i] + global_col
-                    if ds_col < 0 or ds_col >= n_cols:
-                        continue
-                    ds_row = row_offsets[i] + global_row
-                    if ds_row < 0 or ds_row >= n_rows:
+                has_outflow = False
+                for i, ds_col, ds_row, p_ij in (
+                        mfd_flow_direction_raster.yield_downslope_neighbors(
+                            global_col, global_row, skip_oob=False)):
+                    has_outflow = True
+                    if (ds_col < 0 or ds_col >= n_cols or
+                        ds_row < 0 or ds_row >= n_rows):
```

Comment:
The difference is that this code is checking whether the pixel has _any_ outflow, not just outflow to an in-bounds pixel. If `has_outflow` is False, an error is raised below. 

I'm not totally sure if that's necessary - maybe that check should happen in `yield_downslope_neighbors` or somewhere else. But I was trying to preserve the original code structure as much as possible since I don't fully understand it yet